tests/kafkatest/tests/streams/streams_upgrade_test.py

```diff
@@ -38,7 +38,7 @@
 backward_compatible_metadata_2_versions = [str(LATEST_0_10_2), str(LATEST_0_11_0), str(LATEST_1_0), str(LATEST_1_1)]
 # If we add a new version below, we also need to add this version to `streams.py`:
 # -> class `StreamsUpgradeTestJobRunnerService`, method `start_cmd`, variable `KAFKA_STREAMS_VERSION`
-metadata_3_or_higher_versions = [str(LATEST_2_0), str(LATEST_2_1), str(LATEST_2_2), str(LATEST_2_3), str(DEV_VERSION)]
+metadata_3_or_higher_versions = [str(LATEST_2_0), str(LATEST_2_1), str(LATEST_2_2), str(LATEST_2_3), str(LATEST_2_4), str(DEV_VERSION)]
```

Comment:
Hm. We actually changed the metadata version in 2.4, we should have moved `DEV_VERSION` to a new metadata group at that time. I'm not that familiar with the metadata upgrade test, maybe someone else can chime in on what else (if anything) needs to be updated so that we're actually testing the metadata upgrade here?
...I do notice the referenced `start_cmd` function no longer seems to take any versions, so we should update the comment above if anyone knows what's going on here @mjsax @guozhangwang @bbejeck @vvcephei

Reply:
As the variable name says, it's all version with version 3 or higher -- in version 3 we introduced version probing and thus to upgrade, no "upgrade_from" config must be specified (from a metadata version point of view), ie, the metadata version itself can be bumped without an issue.

However, given `2.4` release introduces eager/incremental upgrade, I am not sure if we can simply upgrade from`2.3` to `2.4` without setting the config -- but for a different reason.

Maybe we need to change how we structure the test?

Reply:
I see...I wonder if we should rethink the general approach to upgrade system tests. We have all these different upgrade paths:
a) pre-VP: double rolling bounce, need to pin metadata to lower version
b) version probing: single rolling bounce, but need to wait for group to stabilize
c) cooperative upgrade: double rolling bounce (first of which will technically be VP)

 Obviously it only makes sense to follow/test one path for a given (to, from) pair. Instead of one system test per upgrade path that tests all version combinations, we should have one test that goes through all (to, from) versions and chooses the correct upgrade path for that pair, then just calls it on that pair. WDYT?

note that the current VP system test actually tests forward compatibility, and not any existing upgrade path. We should keep that as a separate test

Reply:
> Instead of one system test per upgrade path that tests all version combinations, we should have one test that goes through all (to, from) versions and chooses the correct upgrade path for that pair, then just calls it on that pair.

Not sure if this would simplify the code? Having "groups of versions" and have a different test for each "upgrade path" (from all versions of one group to all version of the other group) seems to make it clear which "upgrade path" we are testing?

> Note that the current VP system test actually tests forward compatibility, and not any existing upgrade path.

Well, that seems to be fine, as the upgrade path is just a single rolling bounce. But see where you are coming from.

> but it does mean we have a gap in our current system tests w.r.t an actual VP upgrade between existing versions

Why that? If the versions are compatible, there is nothing to be tested but compatibility? Or maybe I miss your point?

Let see what @cadonna thinks. In any case, I don't care too much how we structure the test as long as we don't miss a case :)

Reply:
> Instead of one system test per upgrade path that tests all version combinations, we should have one test that goes through all (to, from) versions and chooses the correct upgrade path for that pair, then just calls it on that pair.

I agree with @mjsax here.

> Note that the current VP system test actually tests forward compatibility, and not any existing upgrade path. That's valuable and we should keep that as a separate test, but it does mean we have a gap in our current system tests w.r.t an actual VP upgrade between existing versions

I am not sure wether I understand all details of `test_version_probing_upgrade`. AFAIU, the test just verifies the version numbers in the subscription and assignment info as well as whether the group stabilizes. That is not a full-fledged real-world upgrade test, but the test rather verifies a fundamental aspect of upgrades with an increased version of metadata. Testing the behavior of the clients according to the changed metadata is done in other system tests like `streams_static_membership_test` and `streams_cooperative_rebalance_upgrade_test`. Please correct me if I am wrong or missed something. I am wondering why we do not test downgrades. Do we not support them?

Anyways, I would not restructure anything in this PR, because this PR should actually only add the 2.4 version to system tests. Since it should also document the update of the system tests for the next release manager, I would like to keep it as simple as possible.

Reply:
I agree that we should keep this simple, and _if_ we want to refactor something do a follow up PR. Also agree, that the purpose of version probing test is, to ensure that we can dump the metadata version without crashing, and it's not a full upgrade test.

> I am wondering why we do not test downgrades. Do we not support them?

Downgrades should work (for some upgrades; not all upgrades are backward compatible). We could add them if we think it's worth the effort. So far, I am not aware of any questions or reported issue and thus I am not sure if it's worth to spent time?

Reply:
I definitely wasn't saying we should block this PR on the issue, refactoring the upgrade tests is way way out of scope here. 
> I don't care too much how we structure the test as long as we don't miss a case :)

I agree completely, but we _have_ missed things: rolling upgrades across a metadata version bump were broken for several versions (this is the "gap" I mentioned -- we test forward compatibility, not backwards, ie we didn't test the 2.0 -> 2.1 upgrade path for example).

We _need_ to test a realistic upgrade scenario; if the existing tests serve other purposes, then maybe we need a new test rather than a refactoring. That said, "test whether we can bump the metadata version" sounds like the job of a unit/integration test, not a system test: the system tests should test an actual upgrade.

I can open a ticket to discuss this further, but I feel pretty strongly that we should reexamine our upgrade-related system tests (I also have strong feelings about what that would look like, but I'm happy to debate that on the ticket). This PR LGTM though, thanks Bruno!